Generate TH functions outside of Type

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20309 Generate TH functions outside of Type**
* #20308 Change view dispatch to abstract

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

I ran a benchmark on pow and before the patch it takes ~3.6ms and after it takes ~3.5ms. I've previously benchmarked the cost of an additional native dispatch to be around 100ns, so this seems as expected.
```
import torch
a = torch.randn(1,1)
%timeit torch.pow(a, 1)
```

Differential Revision: [D15509848](https://our.internmc.facebook.com/intern/diff/D15509848)

Commit:
Wrap TH functions with native functions

colesbury reviewed (changes requested):
What's going on here? Operator overhead has increased 2x since 0.4.1 (and 3x since 0.3.1). This looks like it'll increase it even more and adds a lot of code that does nothing.

This needs:
1) An issue describing what you are trying to achieve
2) Some minimal benchmarks for overhead (i.e. %timeit in ipython before/after on small tensors is fine)

gchanan commented:
I think I know what you are trying to do here, but I'm not 100% sure.  Can you explain in the PR description please?

Commit:
Update on "[wip] Wrap TH functions with native functions"

Wrap TH functions with native functions

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "[wip] Wrap TH functions with native functions"

Wrap TH functions with native functions

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "[wip] Wrap TH functions with native functions"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

ezyang commented:
Since there are code gen changes, it's helpful to reviewers if you post a sample of the generated code before and after.

ezyang commented:
@pytorchbot relabel this please

ezyang commented:
@pytorchbot label this please

ezyang commented on aten/src/ATen/native/LegacyDefinitions.cpp:
```diff
@@ -12,817 +10,957 @@ void* data_ptr(const Tensor & self) {
-Tensor & set_(Tensor& self, Storage source) {
-  return at::legacy::th::_th_set_(self, source);
+Tensor & set__cpu(Tensor& self, Storage source) {
+  return at::legacy::cpu::_th_set_(self, source);
```
Dumb question: Why can't we just have the native functions call the legacy function directly. So for example, instead of `dispatch: CPU: set__cpu`, just say `dispatch: CPU: at::legacy::cpu::_th_set_`?

ezyang replied:
@gchanan, maybe you know the answer to this question?

gchanan replied:
I don't remember if the codegen supports that or not, but it seems like a reasonable thing to try.

ezyang commented on aten/src/ATen/native/cuda/LegacyDefinitions.cu:
```diff
@@ -0,0 +1,930 @@
+#include <ATen/ATen.h>
```
I don't think there is any reason this should be a cu file. And it is a good idea to not put things in cu files if it is not necessary, since nvcc compiles quite a bit more slowly.

gchanan replied:
It doesn't doesn't quite support all of C++.

ezyang commented:
I generally like where this patch is going, and I'd be OK landing it as is under the assumption that it was working (it's not, the ROCm build is failing legitimately:

```
19:40:06 /var/lib/jenkins/workspace/build/aten/src/ATen/LegacyTHFunctionsCUDA.cpp: In function ‘c10::Allocator* at::legacy::cuda::{anonymous}::allocator()’:
19:40:06 /var/lib/jenkins/workspace/build/aten/src/ATen/LegacyTHFunctionsCUDA.cpp:41:21: error: ‘getCUDADeviceAllocator’ is not a member of ‘c10::hip’
19:40:06      return at::hip::getCUDADeviceAllocator();
19:40:06                      ^~~~~~~~~~~~~~~~~~~~~~
19:40:06 /var/lib/jenkins/workspace/build/aten/src/ATen/LegacyTHFunctionsCUDA.cpp:41:21: note: suggested alternative:
19:40:06 In file included from /var/lib/jenkins/workspace/build/aten/src/ATen/LegacyTHFunctionsCUDA.cpp:18:0:
19:40:06 /var/lib/jenkins/workspace/aten/src/ATen/hip/HIPContext.h:61:23: note:   ‘at::cuda::getCUDADeviceAllocator’
19:40:06  CAFFE2_API Allocator* getCUDADeviceAllocator();
19:40:06                        ^~~~~~~~~~~~~~~~~~~~~~
```

Probably you need to replace at namespace with c10 so the regex matches.)

However, I feel with a little more juice in the code generation, we could substantially reduce the boilerplate introduced by this PR. The key observation is that, to hew with the current "style" of how native functions are defined, you had to define a bit pile of new wrapper functions in LegacyDefinitions.cpp (and it's CUDA equivalent). For most of these, you are just doing a direct pass through call to the legacy function. So why not just define the native function to dispatch straight to the legacy function? You'll need to extend the code gen to support dispatching to functions with namespaces explicitly specified, but as far as I can tell, that would not be too much work. Then you can eliminate most of these stubs (not all of them; some of them have special logic, so they'd be fine to keep.) I'm not sure why we didn't do this to start, maybe @gchanan (who introduced the legacy definitions file originally) can say better.

li-roy commented:
@ezyang The body of the generated functions are the same, they are just standalone now.

```
Tensor & CPUType::_th_set_(Tensor & self, Storage source) const;
```
vs
```
namespace at { namespace cpu {
Tensor & _th_set_(Tensor & self, Storage source);
}}
```

Generating the dispatch functions directly was my original plan, but I discussed @gchanan and he preferred explicitly defining them. I don't mind either way, I think both are reasonable.

ezyang commented:
> Generating the dispatch functions directly was my original plan, but I discussed @gchanan and he preferred explicitly defining them. I don't mind either way, I think both are reasonable.

I think I am suggesting something slightly different. My reading of what you said here, was that you were thinking of modifying the code generator so that it directly produced LegacyDefinitions.cpp. I agree with Gregory that an explicit definition is preferable here.

My suggestion is that, in the simple fallthrough cases, you don't need LegacyDefinitions.cpp at all. Instead of specifying that the CPU implementation of `zero_` should call `zero__cpu`, which in turn calls `at::legacy::cpu::_th_zero_`, just say in the dispatch metadata that it should call `at::legacy::cpu::_th_zero_` directly. I may be missing something, but I don't see why this wouldn't work. You always have the option of writing out a stub if there is some extra goop you need for the function in question.

li-roy commented:
Sorry that was a bit unclear. My original plan was to skip the stubs in LegacyDefinitions.cpp (like you describe in your second paragraph). But it's also possible that there was a miscommunication when we first discussed it.

If I modify the codegen, I'll probably choose a namespace like `at::native::legacy::cpu::`, so we can have native as the root namespace.

ezyang commented:
> My original plan was to skip the stubs in LegacyDefinitions.cpp (like you describe in your second paragraph). But it's also possible that there was a miscommunication when we first discussed it.

I suppose, only @gchanan knows :)

gchanan commented:
I think @ezyang's understanding here is correct:
the objection was to autogenerating the definition of the wrapper functions, i.e. building codegen smarts to do that.

Writing the calls directly in native_functions.yaml seems like a good idea, assuming the codegen supports it or can be made to support it easily.

gchanan commented on tools/autograd/derivatives.yaml:
```diff
@@ -1486,3 +1516,47 @@
+- name: multinomial(Tensor self, int64_t num_samples, bool replacement, *, Generator generator)
+  output_differentiability: [False]
+
+- name: is_nonzero(Tensor self)
```
I don't really understand why you need to make the changes to this file.

For example, is_nonzero returns a bool -- how can that possibly be differentiable?

li-roy replied:
I don't think it's required for is_nonzero, i'll remove it. But the others are required because they were initially marked not differentiable because they are BoolTensors or IndexTensors in Declarations.cwrap. Now that we don't have that information, we need to put it here.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -818,10 +818,22 @@
   variants: function, method
 
 - func: fill_(Tensor(a!) self, Scalar value) -> Tensor(a!)
+  cpu_half: True
```
can we do this after https://github.com/pytorch/pytorch/pull/20552?  Pretty ugly otherwise...

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -818,10 +818,22 @@
+  cuda_bool: True
   variants: function, method
+  dispatch:
+    CPU: fill__cpu
```
I know this was mentioned before, but is there a reason we can't do the dispatch directly here?

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1153,7 +1165,7 @@
   variants: function, method
 
-- func: max(Tensor self, int dim, bool keepdim=False, *, Tensor(a!) max, Tensor(b!) max_values) -> (Tensor(a!) values, Tensor(b!) indices)
+- func: max(Tensor self, int dim, bool keepdim=False, *, Tensor(a!) max, Tensor(b!) max_indices) -> (Tensor(a!) values, Tensor(b!) indices)
```
when did this parameter name change happen?  Is this BC breaking?

li-roy replied:
I think it's a typo, but I'll leave it out of this PR

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3358,93 +3980,367 @@
+    CPU: _getri_single_cpu
+    CUDA: _getri_single_cuda
+
+- func: _index_copy_(Tensor(a!) self, int dim, Tensor index, Tensor source) -> Tensor(a!)
```
can we make some more effort to reduce the number of these new functions:
1) all of the linear algebra methods are trivially done, just move the 1d checks into the functions themselves.  It wasn't done that way originally because we couldn't tell the difference between 1-d and 0-d tensors in TH.  This should be a separate PR.
2) min/max/copy_ignoring_overlaps looks like it's only ever used in the context where you know the dispatch device.
3) mode doesn't seem to be used

li-roy replied:
1. I've moved all of them except addr, which I can't do without making error messages worse.
2. min and max must be wrapped because while we only call them when we know it's cuda, it's in a function that serves cpu and cuda. The cuda version doesn't exist when we compile for cpu only, so the dispatch is needed.
3. mode is used in the same file as min and max

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

ezyang commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -286,159 +286,166 @@ RegisterOperators reg({
-     }),
-
-    });
+    DEFINE_TORCH_TENSOR_OP(float, double, at::scalar_to_tensor(scalar_val))
```
What happened in this block?

li-roy replied:
I think it was a bad rebase, this isn't changed in the final commit.

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

ezyang commented:
This looks much better!

ezyang commented:
Waiting on CI to be green for final pass

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

ezyang reviewed (approved):
Aw yeah.

gchanan commented on aten/src/ATen/native/LegacyNNDefinitions.cpp:
```diff
@@ -129,321 +31,15 @@ Tensor nll_loss2d(const Tensor & self, const Tensor & target, const Tensor & wei
 Tensor & log_sigmoid_out(Tensor & output, const Tensor & self) {
   Tensor buffer = at::empty({0}, self.options());
-  return std::get<0>(at::legacy::th::_thnn_log_sigmoid_forward_out(output, buffer, self));
+  return std::get<0>(legacy::cpu::_thnn_log_sigmoid_forward_out(output, buffer, self));
```
how do you know this is should be dispatched to CPU?

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head

Commit:
Update on "Generate TH functions outside of Type"

Generate TH functions outside of Type

This change generates TH functions into at::legacy::cpu and
at::legacy::cuda rather than Type. TH functions will no longer be
visible from Functions.h, NativeFunctions.h, etc. Those interfaces will
now only show native functions.

These new functions in at::legacy are called by native function
wrappers, most of which already exist. In the case of these already
existing wrappers, this saves us a Type dispatch.

Some native function implementions previously called the old generated
TH functions. In these cases, we add a wrapper for those functions and
call it instead. Note the number of Type dispatches stays the same.

We also need to add differentiability logic for some native wrappers
because this was previously done as part of VariableType generation.

gh-metadata: pytorch pytorch 20309 gh/li-roy/14/head
